### ISSUE TITLE
Tests are too slow

### ISSUE THREAD

**Issue by bradfitz** [grpc/grpc-go#534]
The tests are too slow. They can be parallelized, and all sleeps like this should be removed:

``` go
func testNoService(t *testing.T, e env) {
        s, addr := serverSetUp(t, false, nil, math.MaxUint32, nil, nil, e)
        cc := clientSetUp(t, addr, nil, nil, "", e)
        tc := testpb.NewTestServiceClient(cc)
        defer tearDown(s, cc)
        // Make sure setting ack has been sent.                                                                                                                               
        time.Sleep(2 * time.Second)
```

Tests should almost never have sleeps in them. Any sleep duration is always either too slow or too fast (too flaky).

There are always better ways to synchronize.

**Comment by iamqizhao** [comment]
Better ways to sync is definitely preferred and always my favorite when writing a test. All the Sleep remaining in the tests are the ones I have not figured out better ways to synchronize after some thinking.

**Comment by iamqizhao** [comment]
There is no clear action items on this. I close this now. But feel free to send me your optimization PRs. Thanks.

**Comment by bradfitz** [comment]
The actionable thing is to remove all `time.Sleep` calls in tests and add proper synchronization.

Please re-open and assign to me. You can add it to a Low-Priority milestone or label perhaps?

**Comment by dfawley** [comment]
Fixed by #1521.
